client/modules/IDE/pages/IDEView.jsx

```diff
@@ -77,265 +79,243 @@ function WarnIfUnsavedChanges() {
+  const dispatch = useDispatch();
+  const { t } = useTranslation();
 
-    this.state = {
-      consoleSize: props.ide.consoleIsExpanded ? 150 : 29,
-      sidebarSize: props.ide.sidebarIsExpanded ? 160 : 20
-    };
-  }
+  const [consoleSize, setConsoleSize] = useState(
+    ide.consoleIsExpanded ? 150 : 29
+  );
+  const [sidebarSize, setSidebarSize] = useState(
+    ide.sidebarIsExpanded ? 160 : 20
+  );
+  const [isOverlayVisible, setIsOverlayVisible] = useState(true);
 
-  componentDidMount() {
-    // If page doesn't reload after Sign In then we need
-    // to force cleared state to be cleared
-    this.props.clearPersistedState();
+  const rootFile = useSelector(selectRootFile);
+  const canEditProject = useSelector(selectCanEditSketch);
 
-    this.props.stopSketch();
-    if (this.props.params.project_id) {
-      const { project_id: id, username } = this.props.params;
-      if (id !== this.props.project.id) {
-        this.props.getProject(id, username);
-      }
-    }
+  const cmRef = useRef({});
 
-    // window.onbeforeunload = this.handleUnsavedChanges;
-    window.addEventListener('beforeunload', this.handleBeforeUnload);
+  const autosaveIntervalRef = useRef(null);
 
-    this.autosaveInterval = null;
-  }
+  const prevFileNameRef = useRef(selectedFile.name);
+  const prevFileContentRef = useRef(selectedFile.content);
+  const locationRef = useRef(props.location);
 
-  componentWillReceiveProps(nextProps) {
-    if (nextProps.location !== this.props.location) {
-      this.props.setPreviousPath(this.props.location.pathname);
-    }
-    if (this.props.ide.sidebarIsExpanded !== nextProps.ide.sidebarIsExpanded) {
-      this.setState({
-        sidebarSize: nextProps.ide.sidebarIsExpanded ? 160 : 20
-      });
-    }
-  }
+  const syncFileContent = () => {
+    const file = cmRef.current.getContent();
+    dispatch(updateFileContent(file.id, file.content));
+  };
+
+  useEffect(() => {
+    dispatch(clearPersistedState());
 
-  componentWillUpdate(nextProps) {
-    if (nextProps.params.project_id && !this.props.params.project_id) {
-      if (nextProps.params.project_id !== nextProps.project.id) {
-        this.props.getProject(nextProps.params.project_id);
+    dispatch(stopSketch());
+    if (props.params.project_id) {
+      const { project_id: id, username } = props.params;
+      if (id !== project.id) {
+        dispatch(getProject(id, username));
       }
     }
-  }
+  }, []);
+
+  // for setting previous location
+  useEffect(() => {
+    dispatch(setPreviousPath(locationRef.current.pathname));
+  }, [props.location]);
+
+  // for the sidebar size behavior
+  useEffect(() => {
+    if (ide.sidebarIsExpanded) {
+      setSidebarSize((prev) => (prev < 160 ? 160 : sidebarSize));
+    }
+  }, [ide.sidebarIsExpanded]);
 
-  componentDidUpdate(prevProps) {
-    if (this.props.isUserOwner && this.props.project.id) {
+  // For autosave
+  useEffect(() => {
+    let autosaveTimeout;
```

Comment:
I find it confusing having both a local `autosaveTimeout` and a persisted `autosaveIntervalRef`.  It's probably better to remove the `let autosaveTimeout` and replace the usages of `autosaveTimeout` with `autosaveIntervalRef.current`.

I have to look at this bit closer and make sure that it's doing what it's supposed to.  Right now I'm 🤯